src/lib/message-router.js

```diff
@@ -105,7 +106,25 @@ MessageRouter.prototype._handleRequest = function * (request, sender) {
  * @param {Route[]} routes
  * @param {IlpAddress} sender
  */
-MessageRouter.prototype.receiveRoutes = function * (routes, sender) {
+MessageRouter.prototype.receiveRoutes = function * (payload, sender) {
+  debug('receiveRoutes sender:', sender)
+  let routes = payload.new_routes
+
+  let holdDownTime = payload.hold_down_time
+  this.routingTables.bumpConnector(sender, holdDownTime)
+  let potentiallyUnreachableLedgers = payload.unreachable_through_me
+  let lostLedgerLinks = []
+  if (potentiallyUnreachableLedgers.length > 0) {
+    log.info('informed of broken routes to:', potentiallyUnreachableLedgers, ' through:', sender)
+    for (const ledger of potentiallyUnreachableLedgers) {
+      lostLedgerLinks.concat(this.routingTables.invalidateConnectorsRoutesTo(sender, ledger))
+    }
+  }
+
+  if (!routes && (lostLedgerLinks.length === 0)) { // just a heartbeat
+    log.info('got heartbeat from:', sender)
+    return
+  }
   validate('Routes', routes)
```

Comment:
I added schema validation, but the validation is failing and I don't understand why.  The call stack is vague in the extreme, and lots of manual comparison isn't clueing me in:

  1) RouteBroadcaster broadcast invalidates routes:
     InvalidBodyError: RoutingUpdate schema validation error: should be string
      at validate (src/lib/validate.js:18:11)
      at MessageRouter.receiveRoutes (src/lib/message-router.js:111:3)
      at next (native)
      at onFulfilled (node_modules/co/index.js:65:19)
      at node_modules/co/index.js:54:5
      at Context.co (node_modules/co/index.js:50:10)
      at Context.toPromise (node_modules/co/index.js:118:63)
      at next (node_modules/co/index.js:99:29)
      at onFulfilled (node_modules/co/index.js:69:7)
      at node_modules/co/index.js:54:5
      at Context.co (node_modules/co/index.js:50:10)
      at Context.createPromise (node_modules/co/index.js:30:15)
      at Test.Runnable.run (node_modules/co-mocha/lib/co-mocha.js:43:16)

the value that's failing: 
{ new_routes:
   [ { source_ledger: 'eur-ledger.',
       destination_ledger: 'random-ledger.',
       min_message_window: 1,
       source_account: 'eur-ledger.mary',
       points: [Object] } ],
  hold_down_time: 1234,
  unreachable_through_me: [] }